src/library/scala/collection/StringParsers.scala

```diff
@@ -0,0 +1,282 @@
+        //hexDigits.hexDigits 
+        //.hexDigits
+
+        //not .
+
+          if(prefix.charAt(0) == '.') {
+            (prefix != ".") && prefix.substring(1).forall(isHexDigit)
+          } else {
+            val noLeading = prefix.dropWhile(isHexDigit)
+            if(noLeading == "" || noLeading == ".") true
+            else if (noLeading.charAt(0) == '.') noLeading.substring(1).forall(isHexDigit)
+            else false
+          }
+        }
+      }
+
+      def postfixOK(postfix: String) = {
+        if (postfix == "") false
+        else postfix.forall(ch => ch >= '0' && ch <= '9')
+      }
+      // prefix [pP] postfix
+      val pIndex = format.indexWhere(ch => ch == 'p' || ch == 'P')
+      if(pIndex < 1 || pIndex >= format.length - 1){
+        false
+      } 
+      else {
+        val prefix = format.substring(0, pIndex)
+        val postfix = format.substring(pIndex + 1)
+        prefixOK(prefix) && postfixOK(postfix)
+      }
+     }
+
+    def isDecFloatLiteral(format: String): Boolean = {
+      //significant can be one of
+      //* digits.digits
+      //* .digits
+      //* digits.
+      //but not just .
+      if(format == "." || format.startsWith(".e") || format.startsWith(".E") || format.startsWith("E") || format.startsWith("e")) false
+      else {
+        val noSignficant = {
+          val noIntPart = format.dropWhile(ch => ch >= '0' && ch <= '9')
+          val noDecSeparator = if(noIntPart.startsWith(".")) noIntPart.substring(1) else noIntPart
+          noDecSeparator.dropWhile(ch => ch >= '0' && ch <= '9')
+        }
+        if(noSignficant == "") true //this sneaks in just digits. That's OK: It's the "SignedInteger" case
+        else if(noSignficant.startsWith("e") || noSignficant.startsWith("E")) {
+          val exponent = noSignficant.substring(1)
+          if(exponent == "") false
+          else {
+            val unsigned = {
+              val startchar = exponent.charAt(0)
+              if (startchar == '-' || startchar == '+') exponent.substring(1) else exponent
+            }
+            if (unsigned == "") false
+            else unsigned.forall(ch => ch >= '0' && ch <= '9')
+          }
+        } else false
+      }
+    }
+
+    val spacetrimmed = format.trim
+    if (spacetrimmed == "") false
+    else {
+      //all formats can have a sign
+      val unsigned = {
+        val startchar = spacetrimmed.charAt(0)
+        if (startchar == '-' || startchar == '+') spacetrimmed.substring(1) else spacetrimmed
+      }
+      if(unsigned == "") false
+      //that's it for NaN and Infinity
+      else if(unsigned == "NaN" || unsigned == "Infinity") true 
+      else {
+        //all other formats can have a format suffix
+        val desuffixed = {
+          val endchar = unsigned.charAt(unsigned.length - 1)
+          if (endchar == 'f' || endchar == 'F' || endchar == 'd' || endchar == 'D') unsigned.substring(0, unsigned.length - 1)
+          else unsigned
+        }
```

Comment:
Why do floats and doubles get the suffix analysis, but not (that I can see) longs? 
Is "5L" a valid long? 
Is "0x10" a valid base-16 int? 
Is "0x10L" a valid long?
Is "0777" valid? Is it in base-8 or base-10? What about "-0777" ?

Reply:
@paulp the behaviour of the new methods is in all cases the same as the old `toInt`/`toLong`/`toDouble` etc. Those are the same as the equivalent java methods.

In other words, this implementation is intended to have `x.parseFoo == Try(x.toFoo).toOption == Try(java.lang.Foo.parseFoo(x)).toOption` for all `x != null`, with (much) better runtime performance for the unhappy case.

The floating point conversions accept a different set of formats. Why the original distinction was made I don't know exactly (does IEEE 754 or related include a notation guide? If so, probably that, but that's a guess). 

If this were a greenfields implementation, I might be tempted to argue over what trade-offs and formats *I* like, rather than that symmetry (I would, for example, not support any digits outside 7 bit ASCII, which *are* supported for the integral types, but not for the floating point types) - and probably also want to have consistency with scala literals.

But this is not, and I care about  *having* these default parsers to `Option` a great deal more than about which consistencies are the most important ones to support, and which trade-offs are best.

Reply:
That said, while those are the reasons why I think the trade-offs I chose are the most palatable ones, I'm not married to them, and I'm happy to change to any other consensus that you or anyone else is able to find.

Reply:
@martijnhoekstra Thanks. I didn't realize you were reproducing the java behavior in all cases. So am I understanding correctly that you've reproduced all the java library logic solely to avoid the exception? If the reproduced inconsistencies are faithful to the java inconsistencies then that choice seems fine, but it's quite damning for the jvm architecture if the native parsing routines offer a maximum of one element from { performance, correctness }.

Reply:
Yes - the only exception I made is for `null`, which this implementation throws a NPE for (which would escape the Option), and the other implementations handle with a `NumberFormatException` for the integral types (but not for Double and Float, which also throw a NPE for the JDK version).  ¯\\_(ツ)_/¯

This is just failing much faster, and with less typing than `Try("wait, this is not an Int at all!".toInt).toOption`. Such a victory, I know.